R/grade_this.R

```diff
@@ -0,0 +1,70 @@
+
+# #' @export
+# last_result <- function(
+#   env = parent.frame()
+# ) {
+#   if (!exists(".result", envir = env, inherits = TRUE)) {
+#     stop("Could not find `.result`. Be sure to only use `last_result()` inside `grade_this()`")
+#   }
+#   get(".result", envir = env, inherits = TRUE)
+# }
+
+#' @export
+grade_this <- function(expr) {
+  # sometimes expr is a quosure already, so squash all quosures to a single expression
+  express <- rlang::get_expr(rlang::enquo(expr))
+
+  function(checking_env) {
+    eval_gradethis_expr({
+      rlang::eval_tidy(
+        express,
+        env = checking_env
+      )
+    })
+  }
+}
+
+#' @export
+grade_this_code <- function(
+  correct = getOption("gradethis.code.correct", getOption("gradethis.pass", "Correct!")),
+  incorrect = getOption("gradethis.code.incorrect", getOption("gradethis.fail", "Incorrect"))
+) {
+
+  # MUST wrap calling function to be able to shim in `.correct`/`.incorrect`
+  function(checking_env) {
+    checking_env[[".__correct"]] <- correct
+    checking_env[[".__incorrect"]] <- incorrect
+
+    grade_this({
+      # create variable `.message` for glue to find
+      .message <- code_feedback()
+      # call `pass`/`fail` inside `grade_this` to have access to `checking_env`
+      if (is.null(.message)) {
+        # need to use `get()` to avoid using `utils::globalVariables`
+        pass(get(".__correct"))
+      } else {
+        fail(get(".__incorrect"))
+      }
+    })(checking_env)
+  }
+}
+
+
+#' @export
+code_feedback <- function(
```

Comment:
This is the important function, please document well. Context: grading code is no longer considered a best practice, but it is a desirable supplement to every other grading function.